src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java

```diff
@@ -0,0 +1,242 @@
+// limitations under the License.
+package com.google.devtools.build.lib.remote.blobstore.http;
+
+import com.google.auth.Credentials;
+import com.google.devtools.build.lib.remote.blobstore.SimpleBlobStore;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.pool.ChannelPoolHandler;
+import io.netty.channel.pool.SimpleChannelPool;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http.HttpRequestEncoder;
+import io.netty.handler.codec.http.HttpResponseDecoder;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.stream.ChunkedWriteHandler;
+import io.netty.util.internal.PlatformDependent;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLEngine;
+
+/**
+ * Implementation of {@link SimpleBlobStore} that can talk to a HTTP/1.1 backend.
+ *
+ * <p>Blobs (Binary large objects) are uploaded using the {@code PUT} method. Action cache blobs are
+ * stored under the path {@code /ac/base16-key}. CAS (Content Addressable Storage) blobs are stored
+ * under the path {@code /cas/base16-key}. Valid status codes for a successful upload are 200 (OK),
+ * 201 (CREATED), 202 (ACCEPTED) and 204 (NO CONTENT). It's recommended to return 200 (OK) on
+ * success. The other status codes are supported to be compatible with the nginx webdav module and
+ * may be removed in the future.
+ *
+ * <p>Blobs are downloaded using the {@code GET} method at the paths they were stored at. A status
+ * code of 200 should be followed by the content of the blob. The status codes 404 (NOT FOUND) and
+ * 204 (NO CONTENT) indicate that no cache entry exists. It's recommended to return 404 (NOT FOUND)
+ * as the 204 (NO CONTENT) status code is only supported to be compatible with the nginx webdav
+ * module.
+ *
+ * <p>TLS is supported and enabled automatically when using HTTPS as the URI scheme.
+ *
+ * <p>Uploads do not use {@code Expect: 100-CONTINUE} headers, as this would incur an additional
+ * roundtrip for every upload and with little practical value as we would expect most uploads to be
+ * accepted.
+ *
+ * <p>The implementation currently does not support transfer encoding chunked.
+ */
+public class HttpBlobStore implements SimpleBlobStore {
+
+  private final NioEventLoopGroup eventLoop = new NioEventLoopGroup(8);
```

Comment:
What's the "8" here?

Reply:
It's the number of threads. This needs to be documented more clearly, and we may need to be able to inject the number of threads. I'm also a bit concerned about the proliferation of thread-pools in Bazel. I've seen cases where Bazel created several thousand threads on a fairly simple build.

Reply:
done